### DISCUSSION TITLE
Leveraging Neptune read-only endpoint and read/write endpoint

### DISCUSSION THREAD

**Discussion by yongkansage** [Gremlinq/ExRam.Gremlinq#226] [Q&A]
Neptune provides read-only endpoint for read queries, and read/write endpoint for read/write queries. What is the suggested approach to add IGremlinQuerySource service to IServiceCollection container with the two types of endpoints?

**Comment by danielcweber** [comment]
There is, at the time of writing, unfortunately no support for these very Neptune-specific features. As a matter of fact, a Gremlinq query today does not have a notion of whether it's read or write. That might be an interesting concept to implement though - the scenario you proposed would then become possible with some effort (custom executor, etc).

**Reply by yongkansage** [reply]
One workaround but certainly not ideal is having an interface like IGremlinReadonlyQuerySource, and to add its service along with readonly endpoint with UseNeptune to the IServiceCollection during Startup.  It then can be used from DI to deal with read-only type of Gremlin queries later.  Meanwhile, the IGremlinQuerySource service can be used to bound with read/write endpoint, for all the transactions.  Any thoughts?

**Reply by danielcweber** [reply]
I'd rather have a query know by itself whether it's read or write. A custom executor could then leverage this information and choose the right endpoint.

**Reply by yongkansage** [reply]
Can you pls point me to an example usage of custom executor?

**Reply by danielcweber** [reply]
Here's how the executor is set up for generic websockets:
https://github.com/ExRam/ExRam.Gremlinq/blob/626d738ef5065fd60a3c2a86bf7b8b6bf4520c15/src/ExRam.Gremlinq.Providers.WebSocket/Extensions/GremlinQueryEnvironmentExtensions.cs#L247

By UseExecutor/ConfigureExecutor, you can replace/configure any existing executor.

**Comment by danielcweber** [comment]
I created a [short writeup](https://github.com/ExRam/ExRam.Gremlinq/discussions/230) on how to proceed with good ideas. I consider this a good idea.
